### PR TITLE
Add chapter on Optuna for Hyperparameter Optimization

### PR DESCRIPTION
**Pull request by himkt**
This PR tuckles #131.
(Also related to https://github.com/allenai/allennlp/issues/4382)

I create a chapter on Optuna for Hyperparameter Optimization.
I added the tutorial for the basic usage of Optuna integrations for AllenNLP and Optuna functionalities for analyzing the result of optimization.

(This PR is created with squashing, our discussion can be seen [here](https://github.com/himkt/allennlp-guide/pull/1).)

Co-authored with @Crissman and @toshihikoyanase.

### CODE REVIEW

File: exercises/part3/optuna/source.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
+CUDA_DEVICE = 0  # Set device ID if you can use GPU.
```
Comment (himkt) [inline exercises/part3/optuna/source.py:1]: I added the runnable example of optimization.
However, I'm wondering if I should put the link to colab because of the following reasons:

1. Require `optuna`. `binder/requirements.txt` seems to keep small.
2. Require GPU. Hyperparameter optimization takes a longer time than other runnable examples.

I want some feedback about it.

**Comment by matt-gardner** [comment]
Thanks @himkt!  I'm a bit swamped at the moment because of EMNLP review stuff, but I'm hoping to get you a review on this no later than Monday.  Please ping me if I haven't done the review yet by Tuesday.

File: exercises/part3/optuna/source.py (hunk 1 above)
Comment (matt-gardner) [reply exercises/part3/optuna/source.py:1]: If it's not feasible to actually run the code in the example, having it be an executable example isn't really helpful.  So maybe a link to colab is better?

Comment (himkt) [reply exercises/part3/optuna/source.py]: > If it's not feasible to actually run the code in the example, having it be an executable example isn't really helpful.

Agreed. Let me add links to colab this time. eb32f25

**Comment by matt-gardner** [comment]
Sorry @himkt, I was not able to get to this today.  Hopefully tomorrow.

**Comment by himkt** [comment]
@matt-gardner Thank you for letting me know, no problem! 👍

File: chapters/part3/hyperparameter-optimization-with-optuna.md [hunk 2]
```diff
@@ -0,0 +1,426 @@
+---
+title: 'Hyperparameter Optimization with Optuna'
+description: "This chapter gives a basic tutorial for optimizing the hyperparameters of your model."
+type: chapter
```
Comment (matt-gardner) [inline chapters/part3/hyperparameter-optimization-with-optuna.md:4]: You can put an "author" key in here, to have you get credit for writing this (won't render correctly until #134 is merged, though).

File: chapters/part3/hyperparameter-optimization-with-optuna.md [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+`serialization_dir` (directory for saving model snapshot, log, etc.), and `metrics` you want to optimize.
+In the above example, we create an instance of `AllenNLPExecutor` as `executor`.
+Once the `executor` instance is created, training is started with `executor.run()`.
+In each trial step in optimization, the objective function is called and does the following steps:
+
+1. Train a model (`executor.run()`)
+2. Return a target metric on validation data (`executor.run()` returns the specified metric)
+
+Now, we finished defining objective function.
+Let's write Optuna stuff for launching optimization!
```
Comment (matt-gardner) [inline chapters/part3/hyperparameter-optimization-with-optuna.md:272]:
```suggestion
Once we've written an objective function, we can write a script for launching optimization.
```

File: chapters/part3/hyperparameter-optimization-with-optuna.md [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+In Optuna, we create a study object and pass the objective function to the `optimize()` method as follows.
+You can specify something; a way to save a result of optimization, a sampler for searching hyperparameters (`TPESampler` is based on Bayesian Optimization),
+direction for optimizing (maximize or minimize), number of jobs for distributed training, or timeout.
```
Comment (matt-gardner) [inline chapters/part3/hyperparameter-optimization-with-optuna.md:275]:
```suggestion
You can specify something a number of parameters for the hyperparameter optimization process:
- a way to save a result of optimization
- a sampler for searching hyperparameters (`TPESampler` is based on Bayesian Optimization)
- direction for optimizing (maximize or minimize)
- number of jobs for distributed training
- timeout
and more.  An example launch script is below.
```

File: chapters/part3/hyperparameter-optimization-with-optuna.md [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+        n_trials=30,  # number of trials to train a model
+        timeout=timeout,  # threshold for executing time (sec)
+    )
+```
+
+</exercise>
+
+<exercise id="3" title="Results of Hyperparameter Optimization">
+
+If `study.optimize` successfully runs, `trial.db` would be created in the directory `result`.
```
Comment (matt-gardner) [inline chapters/part3/hyperparameter-optimization-with-optuna.md:299]: If the above section is split into two, this one can probably be combined with the second one, to have one exercise talking about running optuna and seeing the results.

File: chapters/part3/hyperparameter-optimization-with-optuna.md [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+
+[Tip] You can change a names of database file and directory by changing the value of `storage` in `optuna.create_study`.
```
Comment (matt-gardner) [inline chapters/part3/hyperparameter-optimization-with-optuna.md:301]: This tip may not be necessary if this text immediately follows the example launch script above, where we already mentioned that we can vary the storage location.

File: chapters/part3/hyperparameter-optimization-with-optuna.md [hunk 7]
```diff
@@ -0,0 +1,426 @@
...
+
+```python
+study.trials_dataframe()
+```
+
+<img src="/part3/hyperparameter-optimization-with-optuna/trials_dataframe.jpg" alt="Dataframe">
+
+Next, visualize a history of optimization.
+To plot a history of optimization, we can use `optuna.visualization.plot_optimization_history`.
+I also put a validation accuracy of baseline model as a reference.
```
Comment (matt-gardner) [inline chapters/part3/hyperparameter-optimization-with-optuna.md:323]:
```suggestion
We also put the validation accuracy of a baseline model as a reference.
```

It's not clear to me how this was done.

File: chapters/part3/hyperparameter-optimization-with-optuna.md [hunk 8]
```diff
@@ -0,0 +1,426 @@
...
+It shows that Optuna successfully found hyperparameters to achieve better performance.
+Note that this figure shows one result of optimization.
+I performed optimization five times with different random seeds and got the average validation accuracy of 0.909 (±0.002), which outperforms the baseline by a large margin.
```
Comment (matt-gardner) [inline chapters/part3/hyperparameter-optimization-with-optuna.md:326]:
```suggestion
We performed optimization five times with different random seeds and got an average validation accuracy of 0.909 (±0.002), which outperforms the baseline by a large margin.
```

File: chapters/part3/hyperparameter-optimization-with-optuna.md [hunk 9]
```diff
@@ -0,0 +1,426 @@
...
+
+```python
+optuna.visualization.plot_optimization_history(study)
+```
+
+<br>
+<img src="/part3/hyperparameter-optimization-with-optuna/optimization_history.jpg" alt="Plot Optimization History" />
+
+From the recent release of Optuna, we can evaluate parameter importances based on finished trials.
```
Comment (matt-gardner) [inline chapters/part3/hyperparameter-optimization-with-optuna.md:335]:
```suggestion
Optuna also lets you evaluate parameter importances based on finished trials.
```

File: chapters/part3/hyperparameter-optimization-with-optuna.md [hunk 10]
```diff
@@ -0,0 +1,426 @@
...
+            model=model,
+            optimizer=optimizer,
+            data_loader=data_loader,
+            validation_data_loader=validation_data_loader,
+            validation_metric="+accuracy",
+            patience=None,
+            num_epochs=50,
+            cuda_device=CUDA_DEVICE,
+            serialization_dir="result",
+            epoch_callbacks=[AllenNLPPruningCallback(trial, "validation_accuracy")],
```
Comment (matt-gardner) [inline chapters/part3/hyperparameter-optimization-with-optuna.md:398]: This is interesting, I didn't know about this.  Is there a way to get the callback instantiated through a config file?  What would have to happen to make that work?  It's pretty annoying to have to instantiate the trainer yourself here, because it messes with a bunch of our typical recommendations for allennlp usage.  If we can make this easier, that would be better.

File: chapters/part3/hyperparameter-optimization-with-optuna.md [hunk 11]
```diff
@@ -0,0 +1,426 @@
...
+description: "This chapter gives a basic tutorial for optimizing the hyperparameters of your model."
+type: chapter
+---
+
+<textblock>
+
+In this chapter, we'll give a quick tutorial of hyperparameter optimization for your AllenNLP model.
+
+</textblock>
+
```
[the remark this answers is not in the document]
Comment (himkt) [reply chapters/part3/hyperparameter-optimization-with-optuna.md]: :+1:

File: chapters/part3/hyperparameter-optimization-with-optuna.md (hunk 6 above)
Comment (himkt) [reply chapters/part3/hyperparameter-optimization-with-optuna.md]: Thank you, I removed the tip since [your suggestion](https://github.com/allenai/allennlp-guide/pull/138#discussion_r472491586) explained this well. b690276

File: chapters/part3/hyperparameter-optimization-with-optuna.md (hunk 2 above)
Comment (himkt) [reply chapters/part3/hyperparameter-optimization-with-optuna.md]: Added in 1119a64. Thank you.

**Comment by himkt** [comment]
Thank you @matt-gardner for the review!
I adopted some of your suggestions/comments.

I'll update the PR to modify the structure of the chapter and clarify the need for algorithms over random search.
Let me ping when I finish revising.

File: chapters/part3/hyperparameter-optimization-with-optuna.md (hunk 10 above)
Comment (matt-gardner) [reply chapters/part3/hyperparameter-optimization-with-optuna.md]: We have a mechanism for passing in arbitrary, in-memory objects when using `FromParams` (see [here](https://guide.allennlp.org/using-config-files#4)).  The main problem is that you need a line somewhere in your code that passes in a `trial` object when you call `from_params`, and you need that value to be passed down to all of your objects recursively.  I tried to figure out a way to do that with minimal modification to our train command, and there isn't a super easy way.  But, here's a bit of a hack that you can do in your own code that abuses the `FromParams` code a bit, but I think it might actually work.

You have [a method](https://github.com/optuna/optuna/blob/9b0081a212cba04e97a28ca510e5e72559b01a0d/optuna/integration/allennlp.py#L144-L145) where you're instantiating a `Params` object.  If that object had a `trial` key with the `Trial` object as a value, it will get passed to `TrainModel.from_params`.  The trouble is that it won't be treated as an `extra`, which it needs to be in order to be passed down recursively to all child classes (like the callbacks that you want to construct).  So, what you can do is create your own subclass of `TrainModel`, registered with a separate name, and have that hijack the `from_params` method and turn the `trial` key from something in the `Params` dictionary to an `extra`.  It would look something like this:

```python
@TrainModel.register("optuna")
class TrainModelWithOptuna(TrainModel):
    def from_params(# match the arguments to `FromParams.from_params`):
        # This is a pretty crazy hack of how FromParams was
        # intended to be used, but it should work for your use case
        trial = params.pop("trial")
        return TrainModel.from_params(params=params, trial=trial, **extras)
```

When objects are passed as extra `kwargs` like this, they are made available to the constructors of every object that gets constructed underneath this object in the code, regardless of whether intermediate classes need that argument (this is how the `Embedding` has access to the `Vocabulary` inside of a `TextFieldEmbedder`, even though the `TextFieldEmbedder` doesn't take a `Vocabulary` as an argument).  So, if we can get `TrainModel.from_params` to get called with `trial=trial`, then the callbacks that you have should get passed the `trial` object correctly.

In order to make this work, where you [instantiate the `Params` object](https://github.com/optuna/optuna/blob/9b0081a212cba04e97a28ca510e5e72559b01a0d/optuna/integration/allennlp.py#L144-L145) you need to add a `"type": "optuna"` key, and a `"trial": trial` key.  The `FromParams` code will call `TrainModel.from_params` with your hacked parameter object, and it will see the type is `optuna`, and call `TrainModelWithOptuna.from_params` after _popping the `type` key_.  Then inside `TrainModelWithOptuna.from_params`, we pop the `trial` key and make it an `extra` when we call `TrainModel.from_params` again.  This time, there is no `type` key, so the default `TrainModel` constructor gets called, but it now has your extra argument.

That should be all that's necessary - only one extra class and a couple of lines of modification to your `run` logic.  Understanding why this works is complicated, though, so I'd definitely recommend having lots of comments in your code explaining what's going on, if you choose to do this.

Comment (himkt) [reply chapters/part3/hyperparameter-optimization-with-optuna.md]: Sorry for the late reply.
I found the way that enables `AllenNLPExecutor` to use `AllenNLPPruningCallback`.

We are working on adding support for Optuna pruning API in https://github.com/optuna/optuna/pull/1772.
After merging this PR, we can easily enable pruning by specifying an epoch callback.
(e.g. https://github.com/himkt/optuna-allennlp/pull/9/files)

1cb7a7a  I updated the content based on the PR.

**Comment by himkt** [comment]
@matt-gardner 
Sorry for the late. I updated the content based on the comments.
Please take a look.

**Review by Crissman — CHANGES_REQUESTED** [review]
Looks good!

File: chapters/part2/hyperparameter-optimization.md [hunk 12]
```diff
@@ -0,0 +1,470 @@
...
+2. Return a target metric on validation data (`executor.run()` returns the specified metric)
+
+Once we've written an objective function, we can write a script for launching optimization.
+In Optuna, we create a study object and pass the objective function to the `optimize()` method as follows.
+You can specify something a number of parameters for the hyperparameter optimization process:
+- a way to save a result of optimization
+- a sampler for searching hyperparameters (`TPESampler` is based on Bayesian Optimization)
+- direction for optimizing (maximize or minimize)
+- number of jobs for distributed training
+- timeout
```
Comment (matt-gardner) [inline chapters/part2/hyperparameter-optimization.md:325]:
```suggestion
- timeout

```

File: chapters/part2/hyperparameter-optimization.md [hunk 13]
```diff
@@ -0,0 +1,470 @@
...
+```
+
+</exercise>
+
+
+<exercise id="6" title="Pruning poor performing trials">
+
+Hyperparameter optimization often takes a long time to find good hyperparameters.
+If you can find and stop unpromising trials with bad hyperparameters, you can reduce the time of optimization and get good hyperparameters faster.
+Stopping unpromising trials is so-called `pruning`.
```
Comment (matt-gardner) [inline chapters/part2/hyperparameter-optimization.md:353]:
```suggestion
Stopping unpromising trials is called `pruning`.
```

File: chapters/part2/hyperparameter-optimization.md [hunk 14]
```diff
@@ -0,0 +1,470 @@
...
+The following illustration shows an example of pruning.
+Although the final curves at the left side of the picture are not available before finishing hyperparameter optimization,
+a pruner evaluates at each epoch how promising each trial will be and stops if it predicts to be unpromising.
```
Comment (matt-gardner) [inline chapters/part2/hyperparameter-optimization.md:356]:
```suggestion
a pruner evaluates at each epoch how promising each trial will be and stops if it predicts it to be unpromising.
```

File: chapters/part2/hyperparameter-optimization.md [hunk 15]
```diff
@@ -0,0 +1,470 @@
...
+
+<img src="/part2/hyperparameter-optimization-with-optuna/illustration_of_pruning.jpg" alt="Illustration of Pruning" />
+
+You can use `AllenNLPPruningCallback` that is the new feature of Optuna,
+which allows users to prune unpromising trials with algorithms implemented in Optuna.
+`AllenNLPPruningCallback` is the interface to provide a way to use these pruning algorithms.
```
Comment (matt-gardner) [inline chapters/part2/hyperparameter-optimization.md:362]:
```suggestion
Optuna provides an `AllenNLPPruningCallback`
which allows users to prune unpromising trials with algorithms implemented in Optuna.
```

File: chapters/part2/hyperparameter-optimization.md [hunk 16]
```diff
@@ -0,0 +1,470 @@
...
+
+You can enable a pruning callback by adding `optuner_pruner` to `epoch_callbacks` in your jsonnet configuration.
```
Comment (matt-gardner) [inline chapters/part2/hyperparameter-optimization.md:364]:
```suggestion
You can enable a pruning callback by adding `optuner_pruner` to `epoch_callbacks` in
your jsonnet configuration (inside the `trainer` parameters).
```

File: chapters/part2/hyperparameter-optimization.md [hunk 17]
```diff
@@ -0,0 +1,470 @@
...
+```json
+   epoch_callbacks: [
+    {
+      type: 'optuna_pruner',
+    },
+  ],
+```
+
+After enabling pruning callback, you have to specify a pruner you want to use.
+In Optuna, some effient algorithms such as [SuccessiveHaving](https://arxiv.org/abs/1502.07943)
```
Comment (matt-gardner) [inline chapters/part2/hyperparameter-optimization.md:377]:
```suggestion
In Optuna, some effient algorithms such as [SuccessiveHalving](https://arxiv.org/abs/1502.07943)
```

**Review by matt-gardner — APPROVED** [review]
This is great!  I left minor copy-editing suggestions, which I will commit myself and then merge this.  Thanks!

File: chapters/part2/hyperparameter-optimization.md (hunk 17 above)
Comment (himkt) [reply chapters/part2/hyperparameter-optimization.md]: Oh, I'm really sorry... thank you for catching.

**Comment by himkt** [comment]
Thank you very much for kindful reviews (and modifications!)!
